src/client/__test__/realUserMonitoring.spec.js

```diff
@@ -0,0 +1,155 @@
+
+// Mock the utility to get a spoor ID from the cookie
+jest.mock('../utils/getSpoorId', () => ({
+	getSpoorId: jest.fn().mockReturnValue('mock-spoor-id')
+}));
+
+// Mock the broadcast method which fires an o-tracking event
+jest.mock('../broadcast', () => ({
+	broadcast: jest.fn()
+}));
+import {broadcast} from '../broadcast';
+
+// Mock Perfume
+jest.mock('perfume.js', () => {
+	return jest.fn();
+});
+import Perfume from 'perfume.js';
+
+// Mock global performance metrics
+Performance.prototype.getEntriesByType = jest.fn().mockReturnValue([
+	{
+		type: 'navigate',
+		domComplete: 123.45,
+		domInteractive: 678.90
+	}
+]);
+
+// Mock requestIdleCallback so we can capture metrics
+// sent by Perfume
+window.requestIdleCallback = jest.fn();
+
+import {realUserMonitoringForPerformance} from '../trackers/realUserMonitoringForPerformance';
+
+describe('src/client/trackers/realUserMonitoringForPerformance', () => {
+
+	describe('.realUserMonitoringForPerformance()', () => {
+
+		beforeEach(() => {
+			realUserMonitoringForPerformance();
+		});
+
+		it('creates a Perfume instance with a custom analytics tracker', () => {
+			expect(Perfume).toHaveBeenCalledTimes(1);
+			const perfumeOptions = Perfume.mock.calls[0][0];
+			expect(typeof perfumeOptions).toBe('object');
+			expect(typeof perfumeOptions.analyticsTracker).toBe('function');
+		});
+
+		describe('analyticsTracker(options)', () => {
+			let analyticsTracker;
+			let oldConsole;
+
+			beforeEach(() => {
+				// We mock the console here so that our library's console.logs don't
+				// make it hard to read the test output
+				oldConsole = global.console;
+				global.console = {log: jest.fn()};
+				analyticsTracker = Perfume.mock.calls[0][0].analyticsTracker;
+			});
+
+			afterEach(() => {
+				global.console = oldConsole;
+			});
+
+			describe('with only one type of metric', () => {
+				beforeEach(() => {
+					analyticsTracker({ metricName: 'fid', data: 13.7 });
+				});
+
+				it('does not broadcast an o-tracking event', () => {
+					expect(broadcast).toHaveBeenCalledTimes(0);
+				});
+			});
+
+			describe('when all metrics are sent', () => {
+				beforeEach(() => {
+					analyticsTracker({ metricName: 'fid', data: 13.7 });
+					analyticsTracker({ metricName: 'lcp', data: 13.7 });
+					analyticsTracker({ metricName: 'ttfb', data: 13.7 });
+					analyticsTracker({ metricName: 'fp', data: 13.7 });
+					analyticsTracker({ metricName: 'fcp', data: 13.7 });
+					analyticsTracker({ metricName: 'cls', data: 13.7 });
+					analyticsTracker({ metricName: 'tbt', data: 13.7 });
+				});
+
+				it('broadcasts an o-tracking event with the formatted metrics data', () => {
+					expect(broadcast).toHaveBeenCalledTimes(1);
+					const broadcastArguments = broadcast.mock.calls[0];
+					expect(broadcastArguments[0]).toBe('oTracking.event');
+					expect(broadcastArguments[1]).toMatchSnapshot();
+				});
+
+				describe('when all metrics are sent a second time', () => {
```

Comment:
**todo:** We only need to send any one of the metrics a second time to test the following line: https://github.com/Financial-Times/n-tracking/blob/7e0cffef82d3a7c13552121d86610f84f68118ac/src/client/trackers/realUserMonitoringForPerformance.js#L85

Because of this we should make the test description and setup clear that we are testing if any one of these metrics is sent again, not if all of the metrics are sent again.

```suggestion
				describe('when any one of the metrics are sent a second time', () => {
```

Reply:
**suggestion:** We could also assert that the `realUserMonitoringForPerformance().context[metric]` value has not been updated.

Reply:
Agreed on only testing one metric a second time and updating the text (both done). I think because the `context` object is never directly exposed/returned by `realUserMonitoringForPerformance` we probably don't need to test it?

The only time that the properties of `context` are used is when the o-tracking event is broadcast, and this test verifies that the event can only be broadcast once so it's kind of irrelevant to the outside world if the `context` values change

Reply:
Ahh yes `context` isn't returned so there is no way to assert it. I now understand that as long as the broadcast is not called again, we don't mind if the context somehow changed as it will have no effect on anything else.